crux/src/Crux/Goal.hs

```diff
@@ -129,24 +135,121 @@ proveToGoal _ allAsmps p pr =
+-- and would in principle enable parallel goal evaluation (though the tree
+-- structure makes that a bit trickier, it isn't too hard).
+--
+-- Note that this function uses the same symbolic backend ('ExprBuilder') as the
+-- symbolic execution phase, which should not be a problem.
+proveGoalsOffline :: forall st sym p asmp ast t fs
+                   . (?outputConfig :: OutputConfig, sym ~ ExprBuilder t st fs)
+                  => WS.SolverAdapter st
+                  -> CruxOptions
+                  -> SimCtxt sym p
+                  -> Maybe (Goals (LPred sym asmp) (LPred sym ast))
+                  -> IO (Maybe (Goals (LPred sym asmp) (LPred sym ast, ProofResult (Either (LPred sym asmp) (LPred sym ast)))))
+proveGoalsOffline _adapter _opts _ctx Nothing = return Nothing
+proveGoalsOffline adapter opts ctx (Just gs0) = do
+  goalNum <- newIORef (0, 0, 0)
+  unless hasUnsatCores $ do
+    sayWarn "Crux" "Warning: Skipping unsat cores because MC-SAT is enabled."
+  Just <$> go goalNum [] gs0
+  where
+    (start,end)
+      | view quiet ?outputConfig = (\_ -> return (), return ())
+      | otherwise = prepStatus "Checking: " (countGoals gs0)
+
+    withTimeout action
+      | Just seconds <- goalTimeout opts = ST.timeout (round seconds * 1000000) action
+      | otherwise = Just <$> action
+
+    hasUnsatCores = not (yicesMCSat opts)
+    failfast = proofGoalsFailFast opts
+
+    go :: IORef (Integer, Integer, Integer)
+       -> [Seq.Seq (LPred sym asmp)]
+       -> Goals (LPred sym asmp) (LPred sym ast)
+       -> IO (Goals (LPred sym asmp) (LPred sym ast, ProofResult (Either (LPred sym asmp) (LPred sym ast))))
+    go goalNum assumptionsInScope gs =
+      case gs of
+        Assuming ps gs1 -> do
+          res <- go goalNum (ps : assumptionsInScope) gs1
+          return (Assuming ps res)
+
+        ProveConj g1 g2 -> do
+          g1' <- go goalNum assumptionsInScope g1
+          (_, _, s) <- readIORef goalNum
+          if failfast && s > 0
+            then return g1'
+            else ProveConj g1' <$> go goalNum assumptionsInScope g2
+
+        Prove p -> do
+          num <- atomicModifyIORef' goalNum (\(val,y,z) -> ((val + 1,y,z), val))
+          start num
+          -- Conjoin all of the in-scope assumptions, the goal, then negate and
+          -- check sat with the adapter
+          let sym = ctx ^. ctxSymInterface
+          assumptions <- WI.andAllOf sym (L.folded . id) (fmap (^. labeledPred) (mconcat assumptionsInScope))
+          goal <- notPred sym (p ^. labeledPred)
+          -- NOTE: We don't currently provide a method for capturing the output
+          -- sent to offline solvers.  We would probably want a file per goal.
+          let logData = WS.defaultLogData
+          mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
+            case satRes of
+              Unsat _ -> do
+                atomicModifyIORef' goalNum (\(x, u, s) -> ((x, u, s+1), ()))
```

Comment:
I think you're incrementing the wrong counter here, and below

Reply:
Whoops, thanks.  I changed it to use a record with named fields to make this mistake harder to repeat